packages/core/src/modules/connections/DidExchangeProtocol.ts

```diff
@@ -0,0 +1,462 @@
+    })
+    const x25519VerificationMethod = getX25519VerificationMethod({
+      id: uuid(),
+      key: x25519Key,
+      controller: '#id',
+    })
+
+    let mediatorRoutingKey
+    if (routing.routingKeys.length > 0) {
+      const [mediatorPublicKeyBase58] = routing.routingKeys
+      const mediatorEd25519Key = Key.fromPublicKeyBase58(mediatorPublicKeyBase58, KeyType.Ed25519)
+      const mediatorEd25519DidKey = new DidKey(mediatorEd25519Key)
+      const mediatorX25519Key = Key.fromPublicKey(
+        convertPublicKeyToX25519(mediatorEd25519Key.publicKey),
+        KeyType.X25519
+      )
+      // Use ed25519 did:key, which also includes the x25519 key used for didcomm
+      mediatorRoutingKey = `${mediatorEd25519DidKey.did}#${mediatorX25519Key.fingerprint}`
+    }
+
+    // TODO Iterate over all endpoints
+    const [serviceEndpoint] = routing.endpoints
+    const service = new DidCommService({
+      id: '#service-0',
+      // Fixme: can we use relative reference (#id) instead of absolute reference here (did:example:123#id)?
+      // We don't know the did yet
+      // TODO we should perhaps use keyAgreement instead of authentication key in here, then it must be changed also in connection record verkey
+      recipientKeys: [ed25519Key.publicKeyBase58],
+      serviceEndpoint,
+      accept: ['didcomm/aip2;env=rfc19'],
+      // It is important that we encode the routing keys as key references.
+      // So instead of using plain verkeys, we should encode them as did:key dids
+      routingKeys: mediatorRoutingKey ? [mediatorRoutingKey] : [],
+    })
+
+    const didDocument = new DidDocumentBuilder('')
+      .addAuthentication(ed25519VerificationMethod)
+      .addKeyAgreement(x25519VerificationMethod)
+      .addService(service)
+      .build()
+
+    const peerDid = DidPeer.fromDidDocument(didDocument, PeerDidNumAlgo.GenesisDoc)
+
+    const didRecord = new DidRecord({
+      id: peerDid.did,
+      role: DidDocumentRole.Created,
+      // It is important to take the did document from the PeerDid class
+      // as it will have the id property
+      // Should not we also resolve and store document for inline peer did?
+      didDocument: peerDid.numAlgo === PeerDidNumAlgo.GenesisDoc ? peerDid.didDocument : undefined,
+      tags: {
+        // We need to save the recipientKeys, so we can find the associated did
+        // of a key when we receive a message from another connection.
+        recipientKeys: peerDid.didDocument.recipientKeys,
+      },
+    })
+
+    await this.didRepository.save(didRecord)
+    this.logger.debug('Did record created.', didRecord)
+    return peerDid
+  }
+
+  private async createSignedAttachment(didDoc: DidDocument, verkey: string) {
+    const didDocAttach = new Attachment({
+      mimeType: 'application/json',
+      data: new AttachmentData({
+        base64: JsonEncoder.toBase64(didDoc),
+      }),
+    })
+
+    const kid = Key.fromPublicKeyBase58(verkey, KeyType.Ed25519)
```

Comment:
I see I put this wrong in the example I made. 

The kid should be a did:key

```suggestion
    const key = Key.fromPublicKeyBase58(verkey, KeyType.Ed25519)
    const kid = new DidKey(key).did
```